images/virtualization-artifact/pkg/controller/vi/internal/source/object_ref_vd.go

```diff
@@ -0,0 +1,361 @@
+				condition.Reason = vicondition.ProvisioningNotStarted
+				condition.Message = service.CapitalizeFirstLetter(err.Error() + ".")
+				return false, nil
+			case errors.Is(err, service.ErrProvisioningFailed):
+				condition.Status = metav1.ConditionFalse
+				condition.Reason = vicondition.ProvisioningFailed
+				condition.Message = service.CapitalizeFirstLetter(err.Error() + ".")
+				return false, nil
+			default:
+				return false, err
+			}
+		}
+
+		err = ds.importerService.Protect(ctx, pod)
+		if err != nil {
+			return false, err
+		}
+
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = vicondition.Provisioning
+		condition.Message = "Import is in the process of provisioning to DVCR."
+
+		vi.Status.Phase = virtv2.ImageProvisioning
+		vi.Status.Progress = ds.statService.GetProgress(vi.GetUID(), pod, vi.Status.Progress)
+		vi.Status.Target.RegistryURL = ds.statService.GetDVCRImageName(pod)
+
+		log.Info("Provisioning...", "progress", vi.Status.Progress, "pod.phase", pod.Status.Phase)
+	}
+
+	return true, nil
+}
+
+func (ds ObjectRefVirtualDisk) StoreToPVC(ctx context.Context, vi *virtv2.VirtualImage, vdRef *virtv2.VirtualDisk, condition *metav1.Condition) (bool, error) {
+	log, _ := logger.GetDataSourceContext(ctx, objectRefDataSource)
+
+	supgen := supplements.NewGenerator(common.VIShortName, vi.Name, vi.Namespace, vi.UID)
+	dv, err := ds.diskService.GetDataVolume(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	pvc, err := ds.diskService.GetPersistentVolumeClaim(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	switch {
+	case isDiskProvisioningFinished(*condition):
+		log.Info("Disk provisioning finished: clean up")
+
+		setPhaseConditionForFinishedImage(pvc, condition, &vi.Status.Phase, supgen)
+
+		// Protect Ready Disk and underlying PVC.
+		err = ds.diskService.Protect(ctx, vi, nil, pvc)
+		if err != nil {
+			return false, err
+		}
+
+		err = ds.diskService.Unprotect(ctx, dv)
+		if err != nil {
+			return false, err
+		}
+
+		return CleanUpSupplements(ctx, vi, ds)
+	case common.AnyTerminating(dv, pvc):
+		log.Info("Waiting for supplements to be terminated")
+	case dv == nil:
+		log.Info("Start import to PVC")
+
+		vi.Status.Progress = "0%"
+		vi.Status.SourceUID = util.GetPointer(vdRef.GetUID())
+
+		if err != nil {
+			setPhaseConditionToFailed(condition, &vi.Status.Phase, err)
+
+			if errors.Is(err, service.ErrInsufficientPVCSize) {
+				return false, nil
+			}
+
+			return false, err
+		}
+
+		source := &cdiv1.DataVolumeSource{
+			PVC: &cdiv1.DataVolumeSourcePVC{
+				Name:      vdRef.Status.Target.PersistentVolumeClaim,
+				Namespace: vdRef.Namespace,
+			},
+		}
+
+		sc, err := ds.diskService.GetStorageClass(ctx, &ds.storageClassForPVC)
+		if err != nil {
+			setPhaseConditionToFailed(condition, &vi.Status.Phase, err)
+			return false, err
+		}
+
+		err = ds.diskService.StartWithIgnoreWFFC(ctx, *vdRef.Spec.PersistentVolumeClaim.Size, sc.GetName(), source, vi, supgen)
```

Comment:
possible panic `*vdRef.Spec.PersistentVolumeClaim.Size`
I think we should change the vd spec by removing the pointer.

The pointer is only needed to understand that the user has indicated something. But it cannot specify 0, and quantity has the IsZero method. Now we have to use these constructions

```
	if vd.Spec.PersistentVolumeClaim.Size != nil && !vd.Spec.PersistentVolumeClaim.Size.IsZero() {
		...
	}
```


I'm ready to discuss this

Reply:
Good idea. We will implement it later